Broker Performance Improvement when there are 10k+ segments

dongxiaoman opened the pull request:
## Description
<!-- Add a description of your PR here.
A good description should include pointers to an issue or design document, etc.
-->
When there are 10k+ segments present, and the partition-based pruner is present (which should be a popular case), the Broker performance is bottlenecked at broker CPU. This PR improved the performance on that case and thus increased our high QPS performance by 7x~8x.

There is more detailed information for the analysis of call stacks and the reason behind in the google doc. We are now able to easily achieve 3k QPS per hosts based on AWS `m5.2xlarge` hosts ([8vPU, spec](https://instances.vantage.sh/?filter=m5.2x&region=us-west-2&cost_duration=annually&reserved_term=yrTerm3Standard.noUpfront))

## How do we improve performance
The hot CPU stacks show that the bottleneck is when the pruner tries to loop through all segments and see if there is segment match. For 10k+ segments, it is an O(n) loop of 10k+ items.

In this PR we try to build and maintain a reverse lookup map of (partition -> list(segments)) in the pruner. Whenever there is a change in the segment list, we rebuild this lookup map. By this change we are able to do O(1) lookup instead of O(n) loop.

Second optimization is that, EmptySegmentPruner is also trying to duplicate the whole segment list. It is not a significant slow down but it shows up in the hot stack list. In our test avoiding this duplication improved QPS from 60k to 80k

## Improvements for our Data Set with 30k+ segments, partitioned
**Before**
<img width="926" alt="Screen Shot 2021-07-27 at 5 39 36 PM" src="https://user-images.githubusercontent.com/11821736/127939752-fd9a0174-9574-4a1c-9912-34ea4eb0810a.png">
**After** **Updated after latest PR change**
<img width="1742" alt="Screen Shot 2021-08-11 at 9 56 18 AM" src="https://user-images.githubusercontent.com/11821736/129071234-9caf2f57-8f1c-4178-a0a1-2f40836caaeb.png">


## Upgrade Notes
Does this PR prevent a zero down-time upgrade? (Assume upgrade order: Controller, Broker, Server, Minion)
* [ ] Yes (Please label as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR fix a zero-downtime upgrade introduced earlier?
* [ ] Yes (Please label this as **<code>backward-incompat</code>**, and complete the section below on Release Notes)

Does this PR otherwise need attention when creating release notes? Things to consider:
- New configuration options
- Deprecation of configurations
- Signature changes to public methods/interfaces
- New plugins added or old plugins removed
* [ ] Yes (Please label this PR as **<code>release-notes</code>** and complete the section on Release Notes)
## Release Notes
<!-- If you have tagged this as either backward-incompat or release-notes,
you MUST add text here that you would like to see appear in release notes of the
next release. -->
Significantly improved Broker CPU performance when partition-pruner is selected for 10k+ segments.

<!-- If you have a series of commits adding or enabling a feature, then
add this section only in final commit that marks the feature completed.
Refer to earlier release notes to see examples of text.
-->
## Documentation
<!-- If you have introduced a new feature or configuration, please add it to the documentation as well.
See https://docs.pinot.apache.org/developers/developers-and-contributors/update-document
-->

jackjlli commented on pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:
```diff
@@ -61,6 +63,7 @@
   private final ZkHelixPropertyStore<ZNRecord> _propertyStore;
   private final String _segmentZKMetadataPathPrefix;
   private final Map<String, PartitionInfo> _partitionInfoMap = new ConcurrentHashMap<>();
+  private final Map<PartitionInfo, Set<String>> _partitionInfoSegments = new ConcurrentHashMap<>();
```
You may probably need to override the hashCode and equals methods in `PartitionInfo` class since the `PartitionInfo` object is used as a key in a hashmap in your PR.

jackjlli replied:
Oh, nvm. Saw your methods down below.

jackjlli commented on pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/SegmentPrunerFactory.java:
```diff
@@ -84,6 +82,8 @@ private SegmentPrunerFactory() {
         }
       }
     }
+    // Always prune out empty segments first
```
Why move it to the bottom?

dongxiaoman replied:
For this change I am a bit concerned whether this change will break some of the existing features, that we may end up selecting some empty segments and later got pruned out, resulting a failed query .

The move is after the below finding:
In our tests, we found that `EmptySegmentPruner` is trying to create a duplicated copy of all the segment names every time. 
It shows up in the jstack outputs, meaning that copying 30k+ segment names for each query is actually not cheap. Moving it to the bottom may mean better performance, in our test it is the difference of **60k QPS => 80k QPS**
(in the case if there is empty segment around)

dongxiaoman replied:
> For the partition pruning, since in most cases a table will use the same partition function for all segments, we can group the segments in a different way:
> 
> * Map<PartitionFunction, Map<Integer, Set>>: partition function -> partition id -> segments
> 
> When getting a query, we can loop over the partition functions from this map and calculate all the matching segments. Then we loop over the input segments and keep the segments that does not have partition info, or included in the matching segments.

Good suggestion; the major challenge of this approach would be analyzing the `isPartitionMatch(Expression filterExpression, PartitionInfo partitionInfo)` method to pick out the segments properly. In a SQL filter we could have something like `WHERE account_key in ['ac23d', 'e92df'] OR account_key = 'c3f9d'`, without proper rewrite it will be hard to make our recursive selection right.

For now my code has become a targeted optimization mostly for Offline table queries, because the externalView is much more stable.

jackjlli commented on pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:
```diff
@@ -278,5 +287,37 @@ private boolean isPartitionMatch(FilterQueryTree filterQueryTree, PartitionInfo
+  private static class CachedPartitionFunction {
+    final Map<PartitionFunction, Map<String, Integer>> cache = new HashMap<>();
+
+    public int getPartition(PartitionFunction func, String value) {
```
Can you elaborate a bit more on the meaning of the `value` here? Does that mean all the values will be cached in the hashmap and will never get cleaned up?

dongxiaoman replied:
For a partition function, it should be a stable fixed map: for each specific function f(x) and a string input x, we should always get the same partition for that immutable string x.
The only major concern could be the cached map size; if we have too many values in the query we could end up having OOM case.

Let me change to use Guava's CacheBuilder if possible https://guava.dev/releases/17.0/api/docs/com/google/common/cache/CacheBuilder.html

jackjlli commented on pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:
```diff
@@ -278,5 +287,37 @@ private boolean isPartitionMatch(FilterQueryTree filterQueryTree, PartitionInfo
+   * Computing partition values may be CPU intensive, caching the computation result may help
+   */
+  private static class CachedPartitionFunction {
+    final Map<PartitionFunction, Map<String, Integer>> cache = new HashMap<>();
```
It would be great to add a brief javadoc on the meaning of the inner hash map.

dongxiaoman replied:
Added Java doc

Jackie-Jiang commented on pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:
```diff
@@ -163,48 +181,44 @@ public synchronized void refreshSegment(String segment) {
+
+  private Set<String> pruneSegments(java.util.function.BiFunction<PartitionInfo, CachedPartitionFunction,  Boolean> partitionMatchLambda) {
+    Set<String> selectedSegments = new HashSet<>();
+    CachedPartitionFunction cachedPartitionFunction = new CachedPartitionFunction();
```
Why are we creating a cache per query? If we just need to cache the partition id, `Map<PartitionFunction, Map<String, Integer>>` should be enough

dongxiaoman replied:
The Cache/LookupTable here is not for our Segment names. It is for the Value -> PartitionNo in each SQL statement. E.g., if we have something like `WHERE account_key = 'a7cd24e'` and we need to compute `MurmurHash.getPartition('a7cd24e')` repeatedly for each segment. If we have 10k segments, we may end up computing  `MurmurHash.getPartition('a7cd24e')` 10k times. 
Because any value can be passed in SQL statement, it makes sense to create the cache for each SQL

klsince replied:
curious any benefit to keep cache across queries? were partitioning values across queries kinda repetitive?

Jackie-Jiang commented on pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:
```diff
@@ -163,48 +181,44 @@ public synchronized void refreshSegment(String segment) {
-        }
-      }
-      return selectedSegments;
+      return pruneSegments((partitionInfo, cachedPartitionFunction) -> isPartitionMatch(filterExpression,
```
(Critical) This algorithm does not prune based on the input segments, and will ignore the results for other pruners. Also, we need to keep the segments with `null` or `INVALID_PARTITION_INFO` because these segments might not be partitioned

dongxiaoman replied:
Yes you are right. Found out in the failed unit tests. Changed and updated now, but needs another round of performance testing.
Will ping you folks for review after another round of testing is done.

Jackie-Jiang reviewed (commented):
For the partition pruning, since in most cases a table will use the same partition function for all segments, we can group the segments in a different way:
- Map<PartitionFunction, Map<Integer, Set<String>>>: partition function -> partition id -> segments

When getting a query, we can loop over the partition functions from this map and calculate all the matching segments. Then we loop over the input segments and keep the segments that does not have partition info, or included in the matching segments.

dongxiaoman commented on pinot-broker/src/test/java/org/apache/pinot/broker/routing/segmentpruner/SegmentPrunerTest.java:
```diff
@@ -241,17 +242,17 @@ public void testPartitionAwareSegmentPruner(QueryCompiler compiler) {
-    assertEquals(segmentPruner.prune(brokerRequest1, Collections.emptySet()), Collections.emptySet());
-    assertEquals(segmentPruner.prune(brokerRequest2, Collections.emptySet()), Collections.emptySet());
-    assertEquals(segmentPruner.prune(brokerRequest3, Collections.emptySet()), Collections.emptySet());
+    assertEquals(segmentPruner.prune(brokerRequest1, new SelectedSegments(Collections.emptySet(), true)).getSegments(), Collections.emptySet());
```
Lots of changes are due to the signature change.
We can avoid these changes by adding a "VisibleForTest" package private wrapper to segmentPruner.
For now let's keep the changes here

dongxiaoman commented on pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:
```diff
@@ -151,10 +170,18 @@ public synchronized void refreshSegment(String segment) {
+//   * Package-level function to make previous tests happy
+//   */
+//  @VisibleForTesting
+//  SelectedSegments prune(BrokerRequest brokerRequest, Set<String> segments) {
```
@Jackie-Jiang because of the signature change below, the Test has lots of changes.
If we add a function here like the one in the comment, we can avoid lots of changes in Test but this method will be only for testing simplicity.
I will remove this comment if we think the tests code change should be like what we have now

dongxiaoman commented:
@Jackie-Jiang @jackjlli I have updated the code to fix the following:
1. As Jackie pointed out, there are cases when externalView is not updated yet but the SegmentSelector has picked up. We have very good test coverage and it help me found out all those failed cases.
2. To support the above cases, introduced `SelectedSegment` class with sorted Sha256 hash so we can quickly compare whether the Set of segments are a match. It helps our cases, and should benefit most Offline table or real time table are not so dynamically updated.

After the optimization, we have either 8x performance improvements, or 2x if the table is dynamic. The 2x improvement comes from avoiding the frequent `MurmurHash.getPartition('a7df')` call.

Please help review again, thanks!

klsince commented on pinot-segment-spi/src/main/java/org/apache/pinot/segment/spi/partition/MurmurPartitionFunction.java:
```diff
@@ -104,4 +105,18 @@ int murmur2(final byte[] data) {
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || ! (o instanceof MurmurPartitionFunction)) return false;
```
same here. could be simplified a bit

Jackie-Jiang commented on pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:
```diff
@@ -163,48 +190,64 @@ public synchronized void refreshSegment(String segment) {
+    // in this case we need to include the new segments.
+    Set<String> selectedSegments = new HashSet<>();
+    CachedPartitionFunction cachedPartitionFunction = new CachedPartitionFunction();
+    if (segments.hasHash() && segments.getSegmentHash().equals(_allSegments.getSegmentHash())) {
```
`TimeSegmentPruner` is always in front of `PartitionSegmentPruner` if configured, so this if statement won't be `true` if any segment is pruned.
It also adds a lot of complexity to the code as we always need to choose whether to calculate the hash.

Jackie-Jiang commented on pinot-broker/src/main/java/org/apache/pinot/broker/routing/segmentpruner/PartitionSegmentPruner.java:
```diff
@@ -88,8 +95,19 @@ public void init(ExternalView externalView, IdealState idealState, Set<String> o
+    Set<String> segmentsFound = new HashSet<>();
+    _partitionInfoMap.forEach((segmentName, partitionInfo) -> {
+      segmentsFound.add(segmentName);
+      Set<String> segments = _partitionInfoSegments.computeIfAbsent(partitionInfo, (info) -> new HashSet<>());
```
This won't be correct if a segment contains more than 1 partitions. We need to maintain a `Map<PartitionFunction, Map<Integer, Set<String>>>`, where the inner map can be optimized to a `List<Set<String>>`, and the index of the list is the partition number.

Jackie-Jiang reviewed (commented):
Introducing `SelectedSegments` adds lots of complexity, and won't handle the case if other pruner prunes the segment.
I'd suggest still go segment by segment, but we can avoid the partition calculation at query time by pre-calculating the partitions for all the segments.
The logic would be as followings:
1. For a given query, calculate the matching partitions, then look up all the matching segments
2. Loop over the input segments. If the segment match the matching segments from step 1, add it to the output; if not, check if it is included in the `_partitionInfoMap`, and add it to the output if it is not included
